src/Test/Tasty/MGolden.hs

```diff
@@ -0,0 +1,123 @@
+module Test.Tasty.MGolden where
+
+import Data.Char (Char)
+import Data.Foldable (traverse_)
+import Data.Proxy (Proxy(..))
+import Data.Typeable (Typeable)
+import MPrelude
+import System.FilePath (FilePath)
+import System.IO (IO)
+import Test.Tasty
+import Test.Tasty.ConsoleFormat
+import Test.Tasty.Options
+import Test.Tasty.Providers
+
+import qualified Data.Algorithm.Diff as Diff
+import qualified Data.Text           as Text
+import qualified Data.Text.IO        as Text
+import qualified System.IO.Error     as Error
+
+data Mode = RunTest | UpdateExpected
+  deriving stock (Eq, Ord, Typeable, Show)
+
+instance IsOption Mode where
+  defaultValue = RunTest
+  parseValue = \case
+    "test"   -> pure RunTest
+    "update" -> pure UpdateExpected
+    _other   -> empty
+
+  optionName     = pure "update"
+  optionHelp     = pure "Update expected on mismatched example"
+  optionCLParser = flagCLParser empty UpdateExpected
+
+data Golden = Golden
+  { action       :: IO Text
+  , expectedPath :: FilePath
+  }
+  deriving stock Typeable
+
+instance IsTest Golden where
+  run options golden _callback = runGolden golden options
+  testOptions = pure . pure $ Option (Proxy :: Proxy Mode)
+
+goldenTest :: FilePath -> IO Text -> TestTree
+goldenTest expectedPath action = singleTest expectedPath Golden{..}
+
+runGolden :: Golden -> OptionSet -> IO Result
+runGolden golden@Golden{..} options = do
+  actual <- action
+
+  maybe
+    (absentFile golden options actual)
+    (testExpected golden options actual)
+    =<< tryRead expectedPath
+
+absentFile :: Golden -> OptionSet -> Text -> IO Result
+absentFile golden@Golden{..} options actual =
+  if shouldUpdate options
+    then updateExpected golden actual
+    else pure $ testFailed "file is absent"
+
+testExpected :: Golden -> OptionSet -> Text -> Text -> IO Result
+testExpected golden@Golden{..} options actual expected =
+  if expected == actual
+    then pure $ testPassed empty
+    else mismatch options golden expected actual
+
+mismatch :: OptionSet -> Golden -> Text -> Text -> IO Result
+mismatch options golden expected actual =
+  if shouldUpdate options
```

Comment:
Instance of https://github.com/mbj/tasty-mgolden/pull/1/files#r337581816

Reply:
See: https://github.com/mbj/tasty-mgolden/pull/1#discussion_r337779774